### PR TITLE
API Conv prim

### PR DESCRIPTION
**Pull request by MuawizChaudhary**
This PR adds the convolution primitive, which generalizes the scattering convolution operation. This PR  will eventually resolve issue #305. Perhaps we can resolve issue #445 ?

### CODE REVIEW

File: kymatio/scattering1d/conv_primitive.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import torch.nn as nn
+import torch.nn.functional as F
+
+class Conv1dFFTPrimative(nn.Module):
```
Comment (eickenberg) [inline kymatio/scattering1d/conv_primitive.py:4]: primitive

File: kymatio/scattering1d/conv_primitive.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    def __init__(self, backend):
+        super(Conv1dFFTPrimative, self).__init__()
+        self.name = 'torch'
+        self.backend = backend
+
+    def preprocess_signal(self, x, pad_left=None, pad_right=None):
```
Comment (eickenberg) [inline kymatio/scattering1d/conv_primitive.py:10]: I would go with `padding` and then a tuple of paddings or a list of tuples of paddings.

The reason I say this is that this fourier primitive can then eventually become N-D and can be used for all dimensionalities. Or we can inherit the different dimensionalities from it and have the inherited class fix the dimension

File: kymatio/scattering1d/conv_primitive.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        if pad_left is not None:
```
Comment (eickenberg) [inline kymatio/scattering1d/conv_primitive.py:11]: I could fuzz attack this by putting `pad_left=5, pad_right=None` and probably cause an error.
And with `pad_left=None, pad_right=5` an inaction.

File: kymatio/scattering1d/conv_primitive.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import torch.nn as nn
```
Comment (eickenberg) [inline kymatio/scattering1d/conv_primitive.py:1]: Maybe call it `convolution.py` for now or `convolution_1d.py` maybe. Do we duplicate filenames already? I think so. In that case use the first.

File: kymatio/scattering1d/core/scattering1d.py [hunk 5]
```diff
@@ -1,11 +1,11 @@
 # Authors: Mathieu Andreux, Joakim Anden, Edouard Oyallon
 # Scientific Ancestry: Joakim Anden, Mathieu Andreux, Vincent Lostanlen
-
+from ..conv_primitive import Conv1dFFTPrimative, Conv1dSpatialPrimative
 
 def scattering1d(x, pad, unpad, backend, J, psi1, psi2, phi, pad_left=0,
         pad_right=0, ind_start=None, ind_end=None, oversampling=0,
         max_order=2, average=True, size_scattering=(0, 0, 0),
-        vectorize=False, out_type='array'):
+        vectorize=False, out_type='array', conv_prim='fft'):
```
Comment (eickenberg) [inline kymatio/scattering1d/core/scattering1d.py:8]: `convolution_type='fft'`

File: kymatio/scattering1d/conv_primitive.py (hunk 1 above)
Comment (eickenberg) [reply kymatio/scattering1d/conv_primitive.py:4]: Actually I would just call it `Conv1dFFT`

File: kymatio/scattering1d/frontend/torch_frontend.py [hunk 6]
```diff
@@ -19,6 +19,7 @@ def __init__(self, J, shape, Q=1, max_order=2, average=True,
         ScatteringBase1D._instantiate_backend(self, 'kymatio.scattering1d.backend.')
         ScatteringBase1D.build(self)
         ScatteringBase1D.create_filters(self)
+        self.conv_prim = 'fft'
```
Comment (eickenberg) [inline kymatio/scattering1d/frontend/torch_frontend.py:22]: `self.convolution` or `self.conv`

File: kymatio/scattering1d/frontend/torch_frontend.py [hunk 7]
```diff
@@ -27,27 +28,41 @@ def register_filters(self):
         saves those arrays as module's buffers."""
         n = 0
         # prepare for pytorch
+        # TODO: have option for different shaped filter depending on
```
Comment (eickenberg) [inline kymatio/scattering1d/frontend/torch_frontend.py:31]: not sure i understand this comment

File: kymatio/scattering1d/frontend/torch_frontend.py [hunk 8]
```diff
@@ -27,27 +28,41 @@ def register_filters(self):
...
+        # convolution primitive?
         for k in self.phi_f.keys():
             if type(k) != str:
                 # view(-1, 1).repeat(1, 2) because real numbers!
-                self.phi_f[k] = torch.from_numpy(
-                    self.phi_f[k]).float().view(-1, 1)
+                if self.conv_prim == 'spatial':
```
Comment (eickenberg) [inline kymatio/scattering1d/frontend/torch_frontend.py:36]: Can you explain this?

Comment (eickenberg) [reply kymatio/scattering1d/frontend/torch_frontend.py:36]: what is `spatial` -- i didn't see it elsewhere

**Comment by eickenberg** [comment]
Since this is rebased on another PR, it is a bit hard to review like this. Let's merge the rfft PR first, then rebase this on top. Then things will become clearer

**Comment by janden** [comment]
Same thing here. The history includes too many unrelated commits. Either rebase or I will review when the relevant PRs are merged.

File: kymatio/scattering1d/conv_primitive.py (hunk 2 above)
Comment (MuawizChaudhary) [reply kymatio/scattering1d/conv_primitive.py:10]: You are smart.

File: kymatio/scattering1d/frontend/torch_frontend.py (hunk 8 above)
Comment (MuawizChaudhary) [reply kymatio/scattering1d/frontend/torch_frontend.py:36]: `spatial` is where we use the torch `Conv1d` module.

Comment (MuawizChaudhary) [reply kymatio/scattering1d/frontend/torch_frontend.py:36]: I had it in, but then removed it. I guess I should remove this code then?

Comment (eickenberg) [reply kymatio/scattering1d/frontend/torch_frontend.py:36]: OH. I guess we need to find a better name for that

**Comment by MuawizChaudhary** [comment]
What will tests for the convolution primitive look like?

File: kymatio/scattering1d/convolution.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import torch.nn as nn
+import torch.nn.functional as F
+
```
Comment (janden) [inline kymatio/scattering1d/convolution.py:3]: Another blank line.

File: kymatio/scattering1d/convolution.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        self.name = 'torch'
+        self.backend = backend
+
+    def preprocess_signal(self, x, padding=()):
+        if len(padding) != 0:
+            x = self.backend.pad(x, pad_left=padding[0], pad_right=padding[1])
+        x_hat = self.backend.rfft(x)
+        return x_hat
+
+    def convolution(self, x_hat, conv_filter, sampling_factor, domain='', ind_start=None, ind_end=None):
```
Comment (janden) [inline kymatio/scattering1d/convolution.py:16]: `sampling` → `subsampling`

File: kymatio/scattering1d/convolution.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import torch.nn as nn
+import torch.nn.functional as F
+
+class Conv1dFFT(nn.Module):
+    def __init__(self, backend):
+        super(Conv1dFFT, self).__init__()
+        self.name = 'torch'
+        self.backend = backend
+
+    def preprocess_signal(self, x, padding=()):
```
Comment (janden) [inline kymatio/scattering1d/convolution.py:10]: Why not have `padding=None`?

File: kymatio/scattering1d/convolution.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+    def convolution(self, x_hat, conv_filter, sampling_factor, domain='', ind_start=None, ind_end=None):
+        y_c = self.backend.cdgmm(x_hat, conv_filter)
+        y_hat = self.backend.subsample_fourier(y_c, 2**sampling_factor)
+        if domain == 'real':
+            y_r = self.backend.irfft(y_hat)
+        else:
+            y_r = self.backend.ifft(y_hat)
+
+        if ind_start is not None:
```
Comment (janden) [inline kymatio/scattering1d/convolution.py:24]: I'm not a huge fan of this. Ideally, we should *always* pad and unpad. It seems weird to only do it in certain circumstances. We should discuss.

File: kymatio/scattering1d/convolution.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+    def preprocess_signal(self, x, padding=()):
+        if len(padding) != 0:
+            x = self.backend.pad(x, pad_left=padding[0], pad_right=padding[1])
+        x_hat = self.backend.rfft(x)
+        return x_hat
+
+    def convolution(self, x_hat, conv_filter, sampling_factor, domain='', ind_start=None, ind_end=None):
+        y_c = self.backend.cdgmm(x_hat, conv_filter)
+        y_hat = self.backend.subsample_fourier(y_c, 2**sampling_factor)
```
Comment (janden) [inline kymatio/scattering1d/convolution.py:18]: Spaces around `**`.

File: kymatio/scattering1d/convolution.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+        if domain == 'real':
```
Comment (janden) [inline kymatio/scattering1d/convolution.py:19]: Ideally, this should be derived from `x` and `conv_filter`.

File: kymatio/scattering1d/core/scattering1d.py [hunk 15]
```diff
@@ -1,11 +1,11 @@
 # Authors: Mathieu Andreux, Joakim Anden, Edouard Oyallon
 # Scientific Ancestry: Joakim Anden, Mathieu Andreux, Vincent Lostanlen
-
+from ..convolution import Conv1dFFT
 
 def scattering1d(x, pad, unpad, backend, J, psi1, psi2, phi, pad_left=0,
         pad_right=0, ind_start=None, ind_end=None, oversampling=0,
         max_order=2, average=True, size_scattering=(0, 0, 0),
-        vectorize=False, out_type='array'):
+        vectorize=False, out_type='array', convolution_type='fft'):
```
Comment (janden) [inline kymatio/scattering1d/core/scattering1d.py:8]: `conv_type` is shorter and aligns better with `Conv1dFFT`.

File: kymatio/scattering1d/convolution.py [hunk 16]
```diff
@@ -0,0 +1,27 @@
+import torch.nn as nn
+import torch.nn.functional as F
+
+class Conv1dFFT(nn.Module):
```
Comment (janden) [inline kymatio/scattering1d/convolution.py:4]: We capitalize `D` elsewhere, right? On the other hand, we may write FFT as `Fft` to make it easier to read.

File: kymatio/scattering1d/core/scattering1d.py [hunk 17]
```diff
@@ -64,6 +64,7 @@ def scattering1d(x, pad, unpad, backend, J, psi1, psi2, phi, pad_left=0,
     irfft = backend.irfft
     cdgmm = backend.cdgmm
     concatenate = backend.concatenate
+    conv = Conv1dFFT(backend)
```
Comment (janden) [inline kymatio/scattering1d/core/scattering1d.py:67]: Should check the `conv_type` argument, right? If only to error if it's not `'fft'`.

File: kymatio/scattering1d/frontend/torch_frontend.py [hunk 18]
```diff
@@ -27,27 +28,41 @@ def register_filters(self):
...
         n = 0
         # prepare for pytorch
+        # TODO: have option for different shaped filter depending on
+        # convolution primitive?
         for k in self.phi_f.keys():
             if type(k) != str:
                 # view(-1, 1).repeat(1, 2) because real numbers!
-                self.phi_f[k] = torch.from_numpy(
-                    self.phi_f[k]).float().view(-1, 1)
+                if self.convolution == 'spatial':
```
Comment (janden) [inline kymatio/scattering1d/frontend/torch_frontend.py:36]: What's going on here?

File: kymatio/scattering1d/convolution.py (hunk 14 above)
Comment (MuawizChaudhary) [reply kymatio/scattering1d/convolution.py:19]: explain?

File: kymatio/scattering1d/frontend/torch_frontend.py (hunk 18 above)
Comment (MuawizChaudhary) [reply kymatio/scattering1d/frontend/torch_frontend.py:36]: This is code I added to make things work for the spatial convolution.  

I should remove it seeing as we are not adding a spatial convolution.

File: kymatio/scattering1d/convolution.py (hunk 14 above)
Comment (janden) [reply kymatio/scattering1d/convolution.py:19]: If `x` is real and `conv_filter` is Hermitian symmetric (that is, its IFT is real), the result of the convolution is real. Checking `x` is easy (we can do that in the `preprocess_signal` step). Checking the symmetry `conv_filter` is expensive, though.

The way to do it is to preprocess both the signal *and* the filter. The filter preprocessing would take place during the initialization of the scattering object (right after they're calculated). In the case of the FFT convolution, the only “preprocessing” would be to check the symmetry and store that as a flag. For other convolutions (like spatial or overlap–add), we would do other types of processing. The spatial convolution, for example, would compute the IFT and cut the result down so that only the non-zero elements are retained.

Longer-term, once we include padding into the convolution, this filter preprocessing would also include calculating the amount of padding that is needed. Short-term, we should try to include a preprocessing step for the filter as well (if only just to check the symmetry of the filter for now).

File: kymatio/scattering1d/frontend/torch_frontend.py (hunk 18 above)
Comment (janden) [reply kymatio/scattering1d/frontend/torch_frontend.py:36]: Ok. Yeah remove it for now.

File: kymatio/scattering1d/core/scattering1d.py (hunk 17 above)
Comment (MuawizChaudhary) [reply kymatio/scattering1d/core/scattering1d.py:67]: The `Conv1DFFT` should be called at the top of the frontend files, methinks. kinda like how the core is called here.

File: kymatio/scattering1d/convolution.py (hunk 14 above)
Comment (MuawizChaudhary) [reply kymatio/scattering1d/convolution.py:18]: I believe that `x` should always be real. I could add that, however the domain specified is asking "should we be taking the real part of this signal after computing the ifft." One way we could abstract this out would be having a operation that takes the real part of the ifft, so we call that in core whenever we want to get the real part of the tensor. This would be an operation in the general torch/numpy/tf backends. 

this is also what we were doing before R2C too.

Comment (janden) [reply kymatio/scattering1d/convolution.py:18]: > I believe that x should always be real.

Right now, yes, but it's not necessarily always true. In the joint time–frequency transform, for example, you have to convolve complex signals sometimes.

> One way we could abstract this out would be having a operation that takes the real part of the ifft, so we call that in core whenever we want to get the real part of the tensor.

Right, but as you point out, that would basically undo a lot of the R2C stuff, which I think is useful. It would also prevent us from taking advantage of an efficient `irfft` if one ever materializes.

As I said, what we want to do is have another component that preprocesses the filters according to the convolution type we're doing. For the FFT, this just involves determining if the filter is real and computing the Fourier transform, storing both.

File: kymatio/scattering1d/core/scattering1d.py (hunk 17 above)
Comment (janden) [reply kymatio/scattering1d/core/scattering1d.py]: Yes the convolution object needs to be tied to the frontend, not to the core. This way it can be involved in the filter preprocessing as discussed above. It'll then be passed into the core and used there, just like the backend.

File: kymatio/scattering1d/convolution.py [hunk 19]
```diff
@@ -0,0 +1,22 @@
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class Conv1DFFT(nn.Module):
```
Comment (eickenberg) [inline kymatio/scattering1d/convolution.py:5]: if this is framework-bound then it should figure in the name
